torch/csrc/autograd/saved_variable.cpp

```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+          message << ", which is output " << output_nr_
+              << " of " << grad_fn->name() << ",";
+      }
+      message << " is at version " << current_version
+          << "; expected version " << saved_version_ << " instead.";
+      if (!AnomalyMode::is_enabled()) {
+          message << " Hint: enable anomaly detection to find the operation "
+              "that failed to compute its gradient, with torch.autograd."
+              "set_detect_anomaly(True).";
+      }
+      else {
+          message << " Hint: the backtrace further above shows the operation "
+              "that failed to compute its gradient. The variable in question "
+              "was changed in there or anywhere later. Good luck!";
+      }
+      TORCH_CHECK(false, message.str());
     }
-    message << " is at version " << current_version
-        << "; expected version " << saved_version_ << " instead.";
-    if (!AnomalyMode::is_enabled()) {
-        message << " Hint: enable anomaly detection to find the operation "
-            "that failed to compute its gradient, with torch.autograd."
-            "set_detect_anomaly(True).";
-    }
-    else {
-        message << " Hint: the backtrace further above shows the operation "
-            "that failed to compute its gradient. The variable in question "
-            "was changed in there or anywhere later. Good luck!";
-    }
-    TORCH_CHECK(false, message.str());
   }
 
-  // The version counter is correct. If we have the original variable, we simply return it
+  // The version counter is correct.
+  // Additionnally, if we deal with a non-leaf variable, we have its correct grad_fn.
 
-  if (saved_original_) {
+  // If we have the original variable, we simply return it
+  if (!hooks_ && saved_original_) {
     return data_;
   }
 
-  // From now on, we can assume the variable is not a leaf and is an output.
-  // Additionnally, because the variable is not a leaf, we have its grad_fn
-  // (computed above) and need to attach it to the returned tensor.
+  const auto data = hooks_ ? hooks_->call_unpack_hook() : data_;
 
   // NB: saved views are unpacked as normal Variables (not views) even though
   // they still share the same storage. This works only because we never call
   // in-place functions on unpacked variables.
-  Variable var = make_variable(data_, Edge(std::move(grad_fn), output_nr_));
+  Variable var;
+  if (grad_fn) {
+    var = make_variable(data, Edge(std::move(grad_fn), output_nr_));
+  } else {
+    var = make_variable(data, requires_grad_);
+  }
+
   impl::set_version_counter(var, version_counter_);
 
+  // If a Variable is a leaf (no grad_fn saved), and it requires_grad, then we
+  // should have saved the grad accumulator. Even if the Variable is no longer
+  // alive, the accumulator should be kept alive by the references in the
+  // graph.
+  if (requires_grad_ && !var.grad_fn() && grad_accumulator_.expired()) {
+    TORCH_CHECK(false, "No grad accumulator for a saved leaf!");
```

Comment:
Should this be TORCH_INTERNAL_ASSERT, or is it possible for user code to reach this point (in which case we could add a test for this?)

Reply:
This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_ASSERT.